Source/DafnyStandardLibraries/src/Std/Actions/Actions.dfy

```diff
@@ -0,0 +1,620 @@
+      requires eventuallyStopsProof.StopFn() == stop
+      requires forall i <- Consumed() :: i == t
+      reads Repr
+      modifies Repr
+      decreases Repr
+      ensures Valid()
+    //ensures history == old(history) + (n copies of t)/(n - 1 not stop values + stop)
+
+    // Convenience methods for specifications
+
+    ghost method Update(t: T, r: R)
+      reads `history
+      modifies `history
+      ensures history == old(history) + [(t, r)]
+    {
+      history := history + [(t, r)];
+    }
+
+    ghost function Consumed(): seq<T>
+      reads this
+    {
+      Inputs(history)
+    }
+
+    ghost function Produced(): seq<R>
+      reads this
+    {
+      Outputs(history)
+    }
+  }
+
+  method DefaultRepeatUntil<T, R>(a: Action<T, R>, t: T, stop: R -> bool, ghost eventuallyStopsProof: ProducesTerminatedProof<T, R>)
+    requires a.Valid()
+    requires eventuallyStopsProof.Action() == a
+    requires eventuallyStopsProof.FixedInput() == t
+    requires eventuallyStopsProof.StopFn() == stop
+    requires forall i <- a.Consumed() :: i == t
+    reads a.Repr
+    modifies a.Repr
+    ensures a.Valid()
+  {
+    while (true)
+      modifies a.Repr
+      invariant fresh(a.Repr - old(a.Repr))
+      invariant a.Valid()
+      invariant forall i <- a.Consumed() :: i == t
+      decreases eventuallyStopsProof.Remaining()
+    {
+      label beforeInvoke:
+      assert a.Valid();
+      assert a.CanProduce(a.history);
+      eventuallyStopsProof.CanConsumeAll(a.history, t);
+      assert a.CanConsume(a.history, t);
+      var next := a.Invoke(t);
+      var nextV := next;
+      if stop(nextV) {
+        break;
+      }
+      eventuallyStopsProof.InvokeUntilTerminationMetricDecreased@beforeInvoke(next);
+    }
+  }
+
+  // Common action invariants
+
+  function Inputs<T, R>(history: seq<(T, R)>): seq<T> {
+    Seq.Map((e: (T, R)) => e.0, history)
+  }
+
+  function Outputs<T, R>(history: seq<(T, R)>): seq<R> {
+    Seq.Map((e: (T, R)) => e.1, history)
+  }
+
+  // A proof that a given action accepts any T value as input,
+  // independent of history.
+  trait ConsumesAllProof<T, R> {
+    ghost function Action(): Action<T, R>
+
+    lemma CanConsumeAll(history: seq<(T, R)>, next: T)
+      requires Action().CanProduce(history)
+      ensures Action().CanConsume(history, next)
+  }
+
+  ghost predicate OnlyProduces<T, R>(i: Action<T, R>, history: seq<(T, R)>, c: R) {
+    i.CanProduce(history) <==> forall e <- history :: e.1 == c
+  }
+
+  ghost predicate Terminated<T>(s: seq<T>, stop: T -> bool, n: nat) {
+    forall i | 0 <= i < |s| :: n <= i <==> stop(s[i])
+  }
+
+  lemma TerminatedDistributes<T>(left: seq<T>, right: seq<T>, stop: T -> bool, n: nat)
+    requires Terminated(left, stop, |left|)
+    requires Terminated(right, stop, n)
+    ensures Terminated(left + right, stop, |left| + n)
+  {}
+
+  lemma TerminatedUndistributes<T>(left: seq<T>, right: seq<T>, stop: T -> bool, n: nat)
```

Comment:
Instead of `...Distributes` and `...Undistributes`, consider `...Composition` and `...Decomposition`.